src/db/roomstore.ts

```diff
@@ -0,0 +1,347 @@
+            };
+            try {
+                await this.db.Run(`INSERT INTO room_entries VALUES ($id, $matrix, $remote)`, values);
+                log.verbose("Created new entry " + entry.id);
+            } catch (ex) {
+                log.error("Failed to insert room entry", ex);
+                throw Error("Failed to insert room entry");
+            }
+        }
+
+        const matrixId = entry.matrix ? entry.matrix.roomId : null;
+        const remoteId = entry.remote ? entry.remote.roomId : null;
+        const mxIdDifferent = matrixId !== row.matrix_id;
+        const rmIdDifferent = remoteId !== row.remote_id;
+        // Did the room ids change?
+        if (row.id && (mxIdDifferent || rmIdDifferent)) {
+            if (matrixId) {
+                this.entriesMatrixIdCache.delete(matrixId);
+            }
+            const items: string[] = [];
+
+            if (mxIdDifferent) {
+                items.push("matrix_id = $matrixId");
+            }
+
+            if (rmIdDifferent) {
+                items.push("remote_id = $remoteId");
+            }
+
+            await this.db.Run(`UPDATE room_entries SET ${items.join(", ")} WHERE id = $id`,
+                {
+                    id: entry.id,
+                    matrixId: matrixId as string|null,
+                    remoteId: remoteId as string|null,
+                },
+            );
+        }
+
+        // Matrix room doesn't store any data.
+        if (entry.remote) {
+            await this.upsertRoom(entry.remote);
+        }
+    }
+
+    public async getEntriesByMatrixId(matrixId: string): Promise<IRoomStoreEntry[]> {
+        const cached = this.entriesMatrixIdCache.get(matrixId);
+        if (cached && cached.ts + ENTRY_CACHE_LIMETIME > Date.now()) {
+            return cached.e;
+        }
+        const entries = await this.db.All(
+            "SELECT * FROM room_entries WHERE matrix_id = $id", {id: matrixId},
+        );
+        const res: IRoomStoreEntry[] = [];
+        for (const entry of entries) {
+            const remoteId = entry.remote_id as string || "";
+            const row = await this.db.Get(
+                "SELECT * FROM remote_room_data WHERE room_id = $rid",
+                {rid: remoteId},
+            );
+            if (!row) { continue; }
+
+            res.push({
+                id: (entry.id as string),
+                matrix: matrixId ? new MatrixStoreRoom(matrixId) : null,
+                // tslint:disable-next-line no-any
+                remote: remoteId ? new RemoteStoreRoom(remoteId, row as any) : null,
+            });
+        }
+        if (res.length > 0) {
+            this.entriesMatrixIdCache.set(matrixId, {e: res, ts: Date.now()});
+        }
+        return res;
+    }
+
+    public async getEntriesByMatrixIds(matrixIds: string[]): Promise<IRoomStoreEntry[]> {
+        const entries = await this.db.All(
+            `SELECT * FROM room_entries WHERE matrix_id IN ('${matrixIds.join("','")}')`,
```

Comment:
this doesn't look particularly SQL injection safe. One could easily make a homeserver which creates sql-injection-ready room IDs and poof goes travis' DB

Reply:
Oh balls, yes you are right (though I hoped that sqlite3 / pg might be doing minimal escaping here).

We should validate that the matrixId really is an id (although I use aliases briefly in MRH).

Reply:
Or we could create an array of parameters `WHERE matrix_id IN ($1, $2, $3, ....)` and then pass those on